### PR TITLE
BUG: array_equivalent_object with mismatched shapes

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/dtypes/missing.py [hunk 1]
```diff
@@ -565,16 +565,7 @@ def _array_equivalent_object(left: np.ndarray, right: np.ndarray, strict_nan: bo
     if not strict_nan:
         # isna considers NaN and None to be equivalent.
 
-        if left.flags["F_CONTIGUOUS"] and right.flags["F_CONTIGUOUS"]:
-            # we can improve performance by doing a copy-free ravel
-            # e.g. in frame_methods.Equals.time_frame_nonunique_equal
-            #  if we transposed the frames
-            left = left.ravel("K")
```
Comment (mroeschke) [inline pandas/core/dtypes/missing.py:572]: So we expect a perf hit but necessary to fix the bug?

Comment (jbrockmendel) [reply pandas/core/dtypes/missing.py:572]: i actually expect this to improve perf in some cases, found the bug along the way

File: pandas/_libs/lib.pyx [hunk 2]
```diff
@@ -571,25 +572,43 @@ def maybe_booleans_to_slice(ndarray[uint8_t, ndim=1] mask):
...
+        y = <object>(<PyObject**>cnp.PyArray_MultiIter_DATA(mi, 1))[0]
 
         # we are either not equal or both nan
         # I think None == None will be true here
         try:
             if PyArray_Check(x) and PyArray_Check(y):
-                if not array_equivalent_object(x, y):
+                if x.shape != y.shape:
+                    return False
+                if x.dtype != y.dtype:
```
Comment (mroeschke) [inline pandas/_libs/lib.pyx:600]: Do we have a test that hits this case (where the shapes are equal but dtypes are not)?

Comment (jbrockmendel) [reply pandas/_libs/lib.pyx]: good catch.  updated with a test and small change to avoid changing this behavior.

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @jbrockmendel

**Pull request merged**